processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -1197,6 +909,39 @@ long getTimestamp()
     }
 
     @Override
+    public boolean equals(Object o)
+    {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+
+      TimeAndDims that = (TimeAndDims) o;
+
+      if (timestamp != that.timestamp) {
+        return false;
+      }
+      for (int i = 0; i < dims.length; i++) {
```

Comment:
Two TimeAndDims might have `dims` of different length so I think this could cause an out of bounds array access on `that.dims[i]`. This also doesn't check `types` at all but the Comparator for sorted facts does check `types`.

The `equals` impl could do `return dimsComparator.compare(this, that) == 0` although this does some needless name lookups.